src/common/src/array/data_chunk.rs

```diff
@@ -65,44 +65,51 @@ impl DataChunkBuilder {
 #[derive(Clone, Default, PartialEq)]
 pub struct DataChunk {
     columns: Vec<Column>,
-    visibility: Option<Bitmap>,
-    cardinality: usize,
+    vis2: Vis,
+}
+
+#[derive(Clone, PartialEq)]
+enum Vis {
+    Bitmap(Bitmap),
+    Compact(usize), // equivalent to all ones of this size
+}
+
+// Indirectly required by StreamChunk::default(), which is only used in tests.
+// Maybe we should impl Default for StreamChunk directly with #[cfg(test)]
+impl Default for Vis {
+    fn default() -> Self {
+        Vis::Compact(0)
+    }
 }
 
 impl DataChunk {
     pub fn new(columns: Vec<Column>, visibility: Option<Bitmap>) -> Self {
-        let cardinality = if let Some(bitmap) = &visibility {
+        let vis2 = if let Some(bitmap) = visibility {
             // with visibility bitmap
-            let card = bitmap.iter().map(|visible| visible as usize).sum();
             for column in &columns {
                 assert_eq!(bitmap.len(), column.array_ref().len())
             }
-            card
+            Vis::Bitmap(bitmap)
         } else if !columns.is_empty() {
             // without visibility bitmap
             let card = columns.first().unwrap().array_ref().len();
             for column in columns.iter().skip(1) {
                 assert_eq!(card, column.array_ref().len())
             }
-            card
+            Vis::Compact(card)
         } else {
             // no data (dummy)
-            0
+            Vis::default()
```

Comment:
This 2nd unreasonable usage of `Default` is exactly the cause of #2663. There should be no default for `Vis`, `DataChunk` and `StreamDataChunk`...